#MBD-seq_analysis7_effect_on_codon_use.R

#UPLOAD THE DATA
#mbd data
setwd("/Users/grovesdixon/git_Repositories/metaTranscriptomes/working_directory")
load("MBD-seq_Image.R")
source("~/git_Repositories/metaTranscriptomes/scripts/MBD-seq_analysis_functions.R")
library(plotrix)

#-------------------- look at delta RSCU values for CpG codons -----------------------------------
#upload RSCU datasets from the concatenated hi and low expression genes
#these are output from get_RSCU_for_fasta.py
hi.ru = read.table('/Users/grovesdixon/Documents/lab_files/projects/metaTranscriptomes/codonW_analyses/get_RSCUs/top5_RSCU.txt', header = T, na.strings = "NA")
low.ru = read.table('/Users/grovesdixon/Documents/lab_files/projects/metaTranscriptomes/codonW_analyses/get_RSCUs/bottom5_RSCU.txt', header = T, na.strings = "NA")

#reformat the datasets to prepare them for plotting
skip.codons = c('UAA', 'UAG', 'UGA', 'AUG', 'UGG')
hi = hi.ru[, !(names(hi.ru) %in% skip.codons)]
low = low.ru[, !(names(low.ru) %in% skip.codons)]
codons = names(hi)[2:length(names(hi))]
hi = as.numeric(t(hi)[2:nrow(t(hi)), ])
low = as.numeric(t(low)[2:nrow(t(low)), ])
dru = data.frame(hi, low, row.names = codons)
dru$d = as.numeric(dru$hi) - as.numeric(dru$low)


#set up two colors to use for barplots
c1 = gray.colors(2, start = 0.3, end = 0.9, gamma = 2.2, alpha = NULL)[2]
c2 = gray.colors(2, start = 0.3, end = 0.9, gamma = 2.2, alpha = NULL)[1]


#plot delta rscus between CG and and other 'double' CC/GC/GG codons
par(mar = c(5, 4, 4, 2) + 0.1)
head(dru)
dru$codons = rownames(dru)
x = get.gc.categories(dru)
cg = x[x$cpg == 1,] ; other = x[x$cpg == 0,]
cg = cg[order(cg$d),] ; other = other[order(other$d),]
cg$type = c1; other$type = c2
res = rbind(cg, other)
bp = barplot(res$d, col = res$type, main = "Delta RSCU")
lab.xs = bp[,1]
# abline(v = lab.xs)
text(lab.xs, -0.12, labels = res$codons, srt = 60, xpd = T, cex = .75)
# segments(.2, mean(cg$d), 8.4, mean(cg$d), lty = 2, col = 'grey', lwd = 2)
# segments(9.8, mean(other$d), 27.6, mean(other$d), lty = 2, lwd = 2)



#build a similar plot for the concatenated ribosomal sequences

#upload the dataset
#this file was generated concatenated ribosomal sequences from each species
#see codonW walkthrough.txt section 'GET RSCU DATA FOR RIBOSOMAL SEQUENCES FROM ALL SPECIES'
r.ru = read.table("all_species_ribosomal_RSCUs_Rinput.txt")
colnames(r.ru) = c('codons', 'aa', 'ru', 'species')
head(r.ru)
r = r.ru[r.ru$species == 'Amillepora',]
x = get.gc.categories(r)
cg = x[x$cpg == 1,] ; other = x[x$cpg == 0,]
cg = cg[order(cg$ru),] ; other = other[order(other$ru),]
cg$type = c1; other$type = c2
res = rbind(cg, other)
# quartz()
bp = barplot(res$ru, col = res$type, main = 'Ribosomal RSCU')
lab.xs = bp[,1]
# abline(v = lab.xs)
text(lab.xs, -0.4, labels = res$codons, srt = 60, xpd = T, cex = .75)



#build similar plot for CAI values

#upload cai data
#this file contains CAI data generated by codonW for all species
#see codonW_walkthrough.txt section 'RUN CODON W ON ALL SPECIES'
cai = read.table("cai_data.txt")
colnames(cai) = c('codons', 'aa', 'xi', 'wi', 'hi', 'opt', 'species')
head(cai)
a = cai[cai$species == 'Amillepora',]
x = get.gc.categories(a)
cg = x[x$cpg == 1,] ; other = x[x$cpg == 0,]
cg = cg[order(cg$wi),] ; other = other[order(other$wi),]
cg$type = c1; other$type = c2
res = rbind(cg, other)
# quartz()
bp = barplot(res$wi, col = res$type, main = 'CAI')
lab.xs = bp[,1]
# abline(v = lab.xs)
text(lab.xs, -0.25, labels = res$codons, srt = 60, xpd = T, cex = .75)


#compare mean RSCU in ribosomes for cg and gc codons by species

#organize the species for plotting
head(r.ru)
species = levels(r.ru$species)
species = c('Apallida', 'Aelegantissima', 'Nvectensis', 'Spistillata', 'Shystrix', 'Pdamicornis', 'Mauretenra', 'Pcarnosus', 'Pdaedalea', 'Mfaveolata', 'Pstrigosa', 'Mcavernosa', 'Fscutaria', 'Pastreoides', 'Ssiderea', 'Atenuis', 'Apalmata', 'Adigitifera', 'Ahyacinthus', 'Amillepora') #these are put in order so that they appear in the barchart the same way as in the phylogenetic tree figure

#set up empty vectors to store the data
mean.cg = c()
mean.other = c()
se.cg = c()
se.other = c()
p.values = c()

#run through the species and gather the data for each
for (sp in species){
	print(paste('Species = ', sp))
	sub = r.ru[r.ru$species == sp,]
	x = get.gc.categories(sub)
	cg = x[x$cpg == 1,] ; other = x[x$cpg == 0,]
	cg = cg[order(cg$ru),] ; other = other[order(other$ru),]
	cg.ru = cg$ru
	other.ru = other$ru
	mean.cg = append(mean.cg, mean(cg.ru))
	mean.other = append(mean.gc, mean(other.ru))
	se.cg = append(se.cg, std.error(cg.ru))
	se.other = append(se.other, std.error(other.ru))
	p = t.test(cg.ru, other.ru, alternative = 'less')$p.value
	p.values = append(p.values, p)
}

#build the plot
quartz()
par(mar = c(5, 8, 4, 2) + 0.1)
res = as.table(rbind(mean.cg, mean.other))
bp = barplot(res, beside = T, axisnames = F, horiz = T, xlim = c(0, max(mean.other + se.other + 0.4)), col = c(c1, c2))
xs = append(bp[1,], bp[2,])
lab.xs = seq(2, 60, by = 3)
text(-1, y = lab.xs, labels = species, xpd = T, cex = 1)
ys = append(res[1,], res[2,])
errs = append(se.cg, se.gc)
plotCI(ys, xs, uiw = errs, err = 'x', add = T, pch = 26, sfrac = 1.5e-3)
title(xlab = "RSCU")
legend(0, 75, c('CG', 'GC/GG/CC'), xpd = T, fill = c(c1, c2))


#gather data for phylogeny plot showing where 'A' Arginines are optimal (based on codonW) and most used in ribosomes
head(r.ru)

a = r.ru[r.ru$species == 'Amillepora',]
a.a = a[a$aa == 'Arg',]
a.a$cg = 0
a.a[grep('CG', a.a$codons), "cg"] <- '1'
a.arg = a.a[a.a$cg == 0,]
c.arg = a.a[a.a$cg == 1,]
hi.c = max(c.arg$ru)
hier.a = a.arg[a.arg$ru > hi.c,]
res = nrow(hier.a)

hier.counts = c()
for (sp in species){
	sub = r.ru[r.ru$species == sp,]
	a.sub = sub[sub$aa == 'Arg',]
	a.sub$cg = 0
	a.sub[grep('CG', a.sub$codons), "cg"] <- '1'
	a.arg = a.sub[a.sub$cg == 0,]
	c.arg = a.sub[a.sub$cg == 1,]
	hi.c = max(c.arg$ru)
	hier.a = a.arg[a.arg$ru > hi.c,]
	res = nrow(hier.a)
	hier.counts = append(hier.counts, res)
}


#gather the data for species where 'A' arganines are optimal 
opt = read.table("hilo_data.txt")
colnames(opt) = c('codons', 'aa', 'hiuse', 'hiCount', 'lowUse', 'lowCount', 'optimal', 'CGcodon', 'GCcodon', 'all.hi.codons', 'all.low.codons', 'species')
head(opt)
c.opt.counts = c()
a.opt.counts = c()
for (sp in species){
	sub = opt[opt$species == sp,]
	a.sub = sub[sub$aa == 'Arg',]
	a.sub$cg = 0
	a.sub[grep('CG', a.sub$codons), "cg"] <- '1'
	a.arg = a.sub[a.sub$cg == 0,]
	c.arg = a.sub[a.sub$cg == 1,]
	c.opt.counts = append(c.opt.counts, sum(c.arg$optimal))
	a.opt.counts = append(a.opt.counts, sum(a.arg$optimal))
}

arg.results = data.frame(species, hier.counts, a.opt.counts)
write.table(arg.results, "Arginine_usage_data.txt", row.names = F, quote = F)





#---------- testing assumptions for automatic generation of Fop and CAI input files ----------
#upload expression data
sdat = read.table("mean_transcript_abundance.txt", header = T)
edat = merge(dat, sdat, by = 'isogroup')
edat = edat[edat$rlog != 0,]

#show correlation with expression

#fop
r = plot.lm('mbd.score', 'fop', dat, '\n', '\n', '\n', limits = F, xlim, c(.2, .8))
wdat = do.window.plot(1/10, 'rlog', 'fop', edat, 10, '\n', '\n', '\n', loess = F, limits = F)
axis(1); axis(2)

#cai
r = plot.lm('mbd.score', 'cai', dat, '\n', '\n', '\n', limits = F, xlim, c(.2, .8))
wdat = do.window.plot(1/10, 'rlog', 'cai', edat, 10, '\n', '\n', '\n', loess = F, limits = F)
axis(1); axis(2)

#make sure that top 5% biased genes are highly expressed
ax = read.table("/Users/grovesdixon/Documents/lab_files/projects/metaTranscriptomes/codonW_analyses/all_genes/axis1.txt", header = T)
head(ax)
ax1 = merge(iso2seq, ax, by = 'contig')
ax = merge(sdat, ax1, by = 'isogroup')
head(ax)
ax = ax[with(ax, order(axis1)),]
head(ax)
p5 = round(0.05 * nrow(ax), 0)
top5 = na.omit(ax[1:p5,])
bottom5 = na.omit(ax[(nrow(ax) - p5) : nrow(ax),])
rest = na.omit(ax[p5+1 : nrow(ax) - p5,])

#set up the means and std.errors
t = mean(top5$rlog)
st = std.error(top5$rlog)
m = mean(rest$rlog)
sm = std.error(rest$rlog)
b = mean(bottom5$rlog)
sb = std.error(bottom5$rlog)
mns = c(t, m, b)
sterr = c(st, sm, sb)

#make dataframe for boxplots and barchart
top5$cat = 'top5'
rest$cat = 'middle'
bottom5$cat = 'bottom5'
bp = rbind(top5, rest, bottom5)

#plot the figures
boxplot(rlog~cat, data = bp)
barplot(mns, width = 1, space = .1, ylim = c(0, 2.7))
plotCI(c(0.6, 1.7, 2.8), mns, sterr, add = T, pch = 26, lwd = 2)
axis(1, at = c(0.6, 1.7, 2.8), labels = c("top 5%", "middle", "bottom 5%"), tick = F)
title(xlab = "Components of Principal Axis", ylab = "Mean Expression")

#test for correlation between principal axis and expression/effective number of codons

#expression
r = plot.lm('axis1', 'rlog', ax, '\n', '\n', '\n', limits = F, xlim, c(.2, .8))

#effective number of codons
nc1 = read.table("effectiveNumberCodons.txt", header = T)
head(nc1)
nc = merge(nc1, ax, by = 'contig')
head(nc)
r = plot.lm('axis1', 'Nc', nc, '\n', '\n', '\n', limits = F, xlim, c(.2, .8))

r = plot.lm('mbd.score', 'Nc', nc2, '\n', '\n', '\n', limits = T, xlim, c(.2, .8))

#look at enrichment for ribosomal genes
#upload the set of ribosomal isogroups
#these come from doing this: grep GO:0005840 amil_defog_iso2go.tab > ribosomal_isogroups.txt. (see CodonW_Instructions.txt for where to get these)
r = read.table("ribosomal_isogroups.txt", col.names = c('isogroup', 'go'))
r = read.table("glycolytic_isogroups.txt", col.names = c('isogroup', 'go'))


head(r)
head(top5)
nrow(top5)
head(ax)
rax = merge(ax, r, by = 'isogroup')
head(rax)

quartz()
par(mfrow = c(2,1))
xlim = c(-.6, 1)
hist(ax$axis1, xlim = xlim)
hist(rax$axis1, xlim = xlim)


#test for enrichment of ribosomal genes
p5 = round(0.05 * nrow(ax), 0)
top5 = na.omit(ax[1:p5,])
other = na.omit(ax[p5:nrow(ax), ])
nrow(top5)
nrow(others)
top.r = nrow(merge(top5, r, by = 'isogroup'))
other.r = nrow(merge(others, r, by = 'isogroup'))
top.not.r = nrow(top5) - top.r
other.not.r = nrow(other) - other.r

r.counts = c(top.r, other.r)
nr.counts = c(top.not.r, other.not.r)
m = as.table(rbind(r.counts, nr.counts))
colnames(m) = c('top5', 'other'); rownames(m) = c('r', 'not.r')
m
fisher.test(m)







abline(v = mean(ax$axis1))

head(dat)


xlim = c(0.6, 0.9)
hist(dat$cai, xlim = xlim)
rop = merge(dat, r, by = 'isogroup')
hist(rop$cai, xlim = xlim)


nrow(rop)

head(ax)
head(dat)


par(mfrow = c(3, 1))
hist(dat$cai, xlim = xlim)
meth = dat[dat$mbd.score > separator,]
not = dat[dat$mbd.score < separator,]
hist(meth$cai, xlim = xlim)
hist(not$cai, xlim = xlim)





